### PR TITLE
Deepcopy for weightedValidator when refreshing proposer list

### PR DESCRIPTION
**Pull request by laggu**
## Proposed changes

- deep copy for weightedValidator when refreshing proposer list

### Problem
calling istanbul.getSnapshot api for a past block sometimes returns validator information of current block.

### reason
A snapshot is made by copying the previous snapshot.
weightedCouncil is a structure in a snapshot.
weightedCouncil has a validator list and a proposer list. They are lists of weightedValidators.
As weightedCouncil is in a snapshot, it is also copied from the previous weightedCouncil.
but the problem is when weightedCouncil is copied, 
a validator list and a proposer list is not deep copied but shallow copied.
so validator object is just one and used for every block.
It is the reason why the snapshot of a past block has current validator information.

## Types of changes

Please put an x in the boxes related to your change.

- [x] Bugfix
- [ ] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

## Further comments
```
func TestWeightedValidator_Copy(t *testing.T) {
	valSet := makeTestWeightedCouncil(testNonZeroWeights)
	copiedValSet := valSet.Copy().(*weightedCouncil)

	t.Logf("valSet.weight : %v copiedValSet.weight : %v",valSet.validators[0].Weight(), copiedValSet.validators[0].Weight())
	atomic.StoreInt64(&(copiedValSet.validators[0].(*weightedValidator).weight),30)
	t.Logf("valSet.weight : %v copiedValSet.weight : %v",valSet.validators[0].Weight(), copiedValSet.validators[0].Weight())
}
```
```
=== RUN   TestWeightedValidator_Copy
--- PASS: TestWeightedValidator_Copy (0.00s)
    weighted_random_test.go:501: valSet.weight : 1 copiedValSet.weight : 1
    weighted_random_test.go:503: valSet.weight : 30 copiedValSet.weight : 30
PASS
```

### CODE REVIEW

File: consensus/istanbul/validator/weighted.go [hunk 1]
```diff
@@ -82,6 +82,16 @@ func (val *weightedValidator) Weight() int64 {
 	return atomic.LoadInt64(&val.weight)
 }
 
+func (val *weightedValidator) Copy() istanbul.Validator {
```
Comment (ehnuje) [inline consensus/istanbul/validator/weighted.go]: Could you please add a test case for this new function?

File: consensus/istanbul/validator/default.go [hunk 2]
```diff
@@ -62,6 +62,12 @@ func (val *defaultValidator) RewardAddress() common.Address { return common.Addr
 func (val *defaultValidator) VotingPower() uint64           { return 1000 }
 func (val *defaultValidator) Weight() int64                 { return 0 }
 
+func (val *defaultValidator) Copy() istanbul.Validator {
```
Comment (ehnuje) [inline consensus/istanbul/validator/default.go]: Same here.

File: consensus/istanbul/backend/engine.go [hunk 3]
```diff
@@ -373,12 +372,10 @@ func (sb *backend) Finalize(chain consensus.ChainReader, header *types.Header, s
 			logger.Trace(logMsg, "header.Number", header.Number.Uint64(), "node address", sb.address, "rewardbase", header.Rewardbase)
 		}
 
-		if proposer.Weight() != 0 {
```
Comment (ehnuje) [inline consensus/istanbul/backend/engine.go]: Why is this line removed?

File: consensus/istanbul/validator/weighted_random_test.go [hunk 4]
```diff
@@ -128,6 +129,35 @@ func makeTestWeightedCouncil(weights []int64) (valSet *weightedCouncil) {
...
+			1000,
+			10,
+		},
+	}
+	for _, testCase := range testCases {
+		validator := newWeightedValidator(testCase.addr, testCase.reward, testCase.votingpower, testCase.weight)
+		copiedValidator := validator.Copy()
+
+		assert.Equal(t, validator, copiedValidator)
+		assert.NotEqual(t, fmt.Sprintf("%p", &validator), fmt.Sprintf("%p", &copiedValidator))
```
Comment (jeonghonah) [inline consensus/istanbul/validator/weighted_random_test.go]: How about comparing the addresses simply without Sprintf?

**Comment by ehnuje** [comment]
@laggu As we discussed, please leave a comment here. Reopen is your choice =)
